Added tools bucket marker for no-downsample.json

RohitKochhar opened the pull request:
Signed-off-by: Rohit Singh <rohitkochhar@bitgo.com>

<!--
    Keep PR title verbose enough and add prefix telling
    about what components it touches e.g "query:" or ".*:"
-->

<!--
    Don't forget about CHANGELOG!

    Changelog entry format:
    - [#<PR-id>](<PR-URL>) Thanos <Component> ...

    <PR-id> Id of your pull request.
    <PR-URL> URL of your PR such as https://github.com/thanos-io/thanos/pull/<PR-id>
    <Component> Component affected by your changes such as Query, Store, Receive.
-->

* [x] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

## Changes

<!-- Enumerate changes you made -->

## Verification

<!-- How you tested it? How do you know it works? -->

RohitKochhar commented:
Linked Issue: https://github.com/thanos-io/thanos/issues/5944

yeya24 commented on pkg/block/block.go:
```diff
@@ -402,3 +402,34 @@ func MarkForNoCompact(ctx context.Context, logger log.Logger, bkt objstore.Bucke
+	}
+	noDownsampleMark, err := json.Marshal(metadata.NoDownsampleMark{
+		ID:      id,
+		Version: metadata.NoCompactMarkVersion1,
```
Shall we update this?

RohitKochhar replied:
Updated!

yeya24 reviewed (commented):
I think we still need to update the downsample code to exclude blocks that have the marker.

RohitKochhar commented:
@yeya24 I have added in the code that filters the block with that filter. Just split it into two commits to reflect adding the flag to the command itself and then another that reflected adding the logic.

yeya24 reviewed (commented):
Thanks! Generally this makes sense to me. Can we add some tests?

RohitKochhar commented:
Sounds good, I'll get working on that over the next day. Do you have any insight why the PR is failing CI? I looked through the action results, but can't diagnose the issue.

yeya24 commented:
> Do you have any insight why the PR is failing CI? I looked through the action results, but can't diagnose the issue.

This one is a flake I believe. Just retried the workflow and let's see if it works this time.

RohitKochhar commented:
@yeya24, adding tests now. I'm wondering what you think would be the best ways to test this. My current idea is to test the `no-downsample` filter and mark creation, similar to `TestIgnoreDeletionMarkFilter_Filter` in `fetcher_test.go#1073`, but besides for that I don't think that there is much to test here

RohitKochhar commented:
@yeya24 test is written, but I think I'm hitting another flake with CI/CD for https://github.com/thanos-io/thanos/actions/runs/3677367451/jobs/6219224415

RohitKochhar commented:
@yeya24 added the test in a similar format to how no-compact marks are testing, looks like it's passing CI so I think this might be good!

yeya24 commented on CHANGELOG.md:
```diff
@@ -19,6 +19,7 @@ We use *breaking :warning:* to mark changes that are not backward compatible (re
 
 ### Added
 
+- [#5945](https://github.com/thanos-io/thanos/pull/5945) Downsample: Add flag to mark blocks for `no-downsample`, similar to the `no-compact.json` marker.
```
Let's make this clearer. We added the new `no-downsample` marker to skip blocks when downsampling. And we added a new flag to mark blocks in the bucket tools rather than the downsample component.

yeya24 replied:
Can we separate them? We did 2 things so please mention both the tools change and the compact change

RohitKochhar commented:
@yeya24 Fixed the changelog entry

fpetkovski commented on pkg/compact/downsample/downsample.go:
```diff
@@ -734,3 +739,104 @@ func SamplesFromTSDBSamples(samples []tsdbutil.Sample) []sample {
+}
+
+// Filter passes all metas, while gathering no downsample markers.
+func (f *GatherNoDownsampleMarkFilter) Filter(ctx context.Context, metas map[ulid.ULID]*metadata.Meta, synced block.GaugeVec, modified block.GaugeVec) error {
```
I wonder if we need to reimplement the filtering logic just because of a new marker. For the short term It should be fine to have it this way, but maybe you can leave a TODO comment to merge the no compact and no downsample filters?

RohitKochhar replied:
Good idea. I'll add in a TODO and open an issue/PR to make that change.

RohitKochhar commented:
@fpetkovski @yeya24 Looking for an update on this... curious why the CI checks seem to have hung...

yeya24 commented:
@RohitKochhar The changelog is not fixed... Please also mention we will skip downsampling for blocks with the new marker

RohitKochhar commented:
@yeya24 thanks for your quick response, I added the additional information in the change log. doing another push seems to have properly retriggered CI as well.

yeya24 reviewed (approved):
LGTM. Thanks for the contribution! @RohitKochhar

RohitKochhar commented:
Perfect, does anything else have to happen to get this merged?
